pipelines-in-pipelines/pkg/reconciler/pip/piprun_test.go

```diff
@@ -0,0 +1,381 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pip
+
+import (
+	"context"
+	"fmt"
+	"github.com/google/go-cmp/cmp"
+	"github.com/tektoncd/experimental/pipelines-in-pipelines/test"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	ttesting "github.com/tektoncd/pipeline/pkg/reconciler/testing"
+	"github.com/tektoncd/pipeline/test/diff"
+	"github.com/tektoncd/pipeline/test/names"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ktesting "k8s.io/client-go/testing"
+	"k8s.io/client-go/tools/record"
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/configmap"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/reconciler"
+	"strings"
+	"testing"
+	"time"
+)
+
+func getPipController(t *testing.T, d test.Data) (test.Assets, func()) {
+	ctx, _ := ttesting.SetupFakeContext(t)
+	ctx, cancel := context.WithCancel(ctx)
+	c, informers := test.SeedTestData(t, ctx, d)
+
+	configMapWatcher := configmap.NewStaticWatcher()
+	ctl := NewController(ctx, configMapWatcher)
+
+	if la, ok := ctl.Reconciler.(reconciler.LeaderAware); ok {
+		la.Promote(reconciler.UniversalBucket(), func(reconciler.Bucket, types.NamespacedName) {})
+	}
+	if err := configMapWatcher.Start(ctx.Done()); err != nil {
+		t.Fatalf("error starting configmap watcher: %v", err)
+	}
+
+	return test.Assets{
+		Logger:     logging.FromContext(ctx),
+		Controller: ctl,
+		Clients:    c,
+		Informers:  informers,
+		Recorder:   controller.GetEventRecorder(ctx).(*record.FakeRecorder),
+	}, cancel
+}
+
+func checkRunCondition(t *testing.T, run *v1alpha1.Run, expectedStatus corev1.ConditionStatus, expectedReason string, expectedMessage string) {
+	condition := run.Status.GetCondition(apis.ConditionSucceeded)
+	if condition == nil {
+		t.Error("Condition missing in Run")
+	} else {
+		if condition.Status != expectedStatus {
+			t.Errorf("Expected Run status to be %v but was %v", expectedStatus, condition)
+		}
+		if condition.Reason != expectedReason {
+			t.Errorf("Expected reason to be %q but was %q", expectedReason, condition.Reason)
+		}
+		if condition.Message != expectedMessage {
+			t.Errorf("Expected message to be %q but was %q", expectedMessage, condition.Message)
+		}
+	}
+	if run.Status.StartTime == nil {
+		t.Errorf("Expected Run start time to be set but it wasn't")
+	}
+	if expectedStatus == corev1.ConditionUnknown {
+		if run.Status.CompletionTime != nil {
+			t.Errorf("Expected Run completion time to not be set but it was")
+		}
+	} else if run.Status.CompletionTime == nil {
+		t.Errorf("Expected Run completion time to be set but it wasn't")
+	}
+}
+
+func checkEvents(fr *record.FakeRecorder, testName string, wantEvents []string) error {
```

Comment:
This function is getting copied in a few places, I wonder if we should export it from a test packages somewhere... but it's ok to copy it here :)

Reply:
agreed, will need to do that because it's used to check events in all the custom tasks controllers